roll/ast.py

```diff
@@ -0,0 +1,685 @@
+import logging
+import random
+from enum import Enum, auto
+from functools import wraps
+
+import roll.exceptions as rollerr
+from roll.types import *
+from utils.exceptions import OutputTooLargeError
+
+MAX_ROLLS = 1000
+
+
+def isfunction(token):
+    return isinstance(token, TokenFunction)
+
+
+class Environment:
+    def __init__(self, root):
+        self.root = root
+        self.closure = {}
+        self.trace = []
+
+    def copy(self):
+        out = Environment(self.root)
+        out.closure = self.closure.copy()
+        out.trace = self.trace
+        return out
+
+
+class HashCounter:
+    """A mutating object that records the variable IDs to be used in general hashing and eta-reduction hashing"""
+
+    def __init__(self):
+        self.__next_id = 0
+        self.__next_scope_id = -1
+        self.trace = []
+
+    def copy(self):
+        out = HashCounter()
+        out.__next_id = self.__next_id
+        out.__next_scope_id = self.__next_scope_id
+        out.trace = self.trace
+        return out

```

Comment:
This is equivalent to using https://docs.python.org/3/library/copy.html?#copy.copy. Note that `trace` will be the same list in the copy under both the current implementation and the shallow copy above, so changing the list in one will change the list in the other.